File: 2.- Practica/Ejercicio 2/ejercicioDos.py
import heapq
from collections import deque

# Estado objetivo del 15-puzzle
estadoMeta = [
    [1, 2, 3, 4],
    [5, 6, 7, 8],
    [9, 10, 11, 12],
    [13, 14, 15, 0]
]

# Movimientos posibles: (dx, dy)
# (-1,0): arriba, (1,0): abajo, (0,-1): izquierda, (0,1): derecha
movimientosPosibles = [(-1,0), (1,0), (0,-1), (0,1)]

def tableroATupla(tablero):
    """Convierte el tablero (lista de listas) a una tupla de tuplas para hashing."""
    return tuple(tuple(fila) for fila in tablero)

def buscarCero(tablero):
    """Encuentra la posición (fila, columna) del espacio vacío (0) en el tablero."""
    for fila in range(4):
        for columna in range(4):
            if tablero[fila][columna] == 0:
                return fila, columna
    return -1, -1

def vecinos(tablero):
    """Genera los tableros vecinos moviendo el espacio vacío en las direcciones posibles."""
    listaVecinos = []
    filaCero, columnaCero = buscarCero(tablero)
    for movFila, movColumna in movimientosPosibles:
        nuevaFila, nuevaColumna = filaCero + movFila, columnaCero + movColumna
        if 0 <= nuevaFila < 4 and 0 <= nuevaColumna < 4:
            nuevoTablero = [fila[:] for fila in tablero]
            nuevoTablero[filaCero][columnaCero], nuevoTablero[nuevaFila][nuevaColumna] = nuevoTablero[nuevaFila][nuevaColumna], nuevoTablero[filaCero][columnaCero]
            listaVecinos.append(nuevoTablero)
    return listaVecinos

def esMeta(tablero):
    """Verifica si el tablero es igual al estado meta."""
    return tablero == estadoMeta

def manhattan(tablero):
    """Calcula la suma de las distancias Manhattan de cada ficha a su posición objetivo."""
    distanciaTotal = 0
    for fila in range(4):
        for columna in range(4):
            valor = tablero[fila][columna]
            if valor != 0:
                filaMeta = (valor - 1) // 4
                columnaMeta = (valor - 1) % 4
                distanciaTotal += abs(fila - filaMeta) + abs(columna - columnaMeta)
    return distanciaTotal

def busquedaCiegas(tableroInicial):
    """Algoritmo de búsqueda a ciegas (BFS) para encontrar la solución."""
    cola = deque()
    cola.append((tableroInicial, []))
    estadosVisitados = set()
    estadosVisitados.add(tableroATupla(tableroInicial))
    while cola:
        tableroActual, camino = cola.popleft()
        if esMeta(tableroActual):
            return camino + [tableroActual]
        for vecinoTablero in vecinos(tableroActual):
            vecinoTupla = tableroATupla(vecinoTablero)
            if vecinoTupla not in estadosVisitados:
                estadosVisitados.add(vecinoTupla)
                cola.append((vecinoTablero, camino + [tableroActual]))
    return None

def busquedaA(tableroInicial):
    """Algoritmo de búsqueda informada (A*) usando la heurística de Manhattan."""
    monticulo = []
    heapq.heappush(monticulo, (manhattan(tableroInicial), 0, tableroInicial, []))
    estadosVisitados = set()
    estadosVisitados.add(tableroATupla(tableroInicial))
    while monticulo:
        costoEstimado, costoActual, tableroActual, camino = heapq.heappop(monticulo)
        if esMeta(tableroActual):
            return camino + [tableroActual]
        for vecinoTablero in vecinos(tableroActual):
            vecinoTupla = tableroATupla(vecinoTablero)
            if vecinoTupla not in estadosVisitados:
                estadosVisitados.add(vecinoTupla)
                nuevoCosto = costoActual + 1
                heuristica = manhattan(vecinoTablero)
                heapq.heappush(monticulo, (nuevoCosto + heuristica, nuevoCosto, vecinoTablero, camino + [tableroActual]))
    return None

def imprimirTablero(tablero):
    """Imprime el tablero de forma legible."""
    for fila in tablero:
        print(' '.join(str(x).rjust(2) for x in fila))
    print()

if __name__ == "__main__":
    # Estado inicial (puedes modificarlo)
    tableroInicial = [
        [5, 1, 2, 4],
        [0, 6, 3, 8],
        [9, 10, 7, 12],
        [13, 14, 11, 15]
    ]
    print("Solución búsqueda a ciegas (puede ser lenta):")
    solucionCiegas = busquedaCiegas(tableroInicial)
    if solucionCiegas:
        for tablero in solucionCiegas:
            imprimirTablero(tablero)
    else:
        print("No se encontró solución con búsqueda a ciegas.")

    print("\nSolución búsqueda informada (A*):")
    solucionA = busquedaA(tableroInicial)
    if solucionA:
        for tablero in solucionA:
            imprimirTablero(tablero)
    else:
        print("No se encontró solución con búsqueda informada.")

File: 2.- Practica/Ejercicio 2/ejercicioDos.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define N 4
#define MAX_ESTADOS 100000

typedef struct {
    int tablero[N][N];
    int pasos;
    int padre;
    int heuristica;
} Estado;

// Prototipos de funciones
int iguales(int a[N][N], int b[N][N]);
void copiarTablero(int dest[N][N], int src[N][N]);
void buscarCero(int tablero[N][N], int *fila, int *col);
int manhattan(int tablero[N][N]);
void imprimirTablero(int tablero[N][N]);
int busquedaCiegas(int tableroInicial[N][N], Estado estados[MAX_ESTADOS], int *solucionIdx);
int busquedaA(int tableroInicial[N][N], Estado estados[MAX_ESTADOS], int *solucionIdx);
void imprimirSolucion(Estado estados[MAX_ESTADOS], int solucionIdx);

const int estadoMeta[N][N] = {
    {1, 2, 3, 4},
    {5, 6, 7, 8},
    {9, 10, 11, 12},
    {13, 14, 15, 0}
};

const int movimientosPosibles[4][2] = {
    {-1,0}, {1,0}, {0,-1}, {0,1}
};

int iguales(int a[N][N], int b[N][N]) {
    for(int i=0;i<N;i++)
        for(int j=0;j<N;j++)
            if(a[i][j]!=b[i][j]) return 0;
    return 1;
}

void copiarTablero(int dest[N][N], int src[N][N]) {
    for(int i=0;i<N;i++)
        for(int j=0;j<N;j++)
            dest[i][j]=src[i][j];
}

void buscarCero(int tablero[N][N], int *fila, int *col) {
    for(int i=0;i<N;i++)
        for(int j=0;j<N;j++)
            if(tablero[i][j]==0) {
                *fila=i; *col=j; return;
            }
}

int manhattan(int tablero[N][N]) {
    int distanciaTotal=0;
    for(int i=0;i<N;i++)
        for(int j=0;j<N;j++) {
            int valor=tablero[i][j];
            if(valor!=0) {
                int filaMeta=(valor-1)/N;
                int colMeta=(valor-1)%N;
                distanciaTotal+=abs(i-filaMeta)+abs(j-colMeta);
            }
        }
    return distanciaTotal;
}

void imprimirTablero(int tablero[N][N]) {
    for(int i=0;i<N;i++) {
        for(int j=0;j<N;j++)
            printf("%2d ", tablero[i][j]);
        printf("\n");
    }
    printf("\n");
}

// BFS (búsqueda ciega)
int busquedaCiegas(int tableroInicial[N][N], Estado estados[MAX_ESTADOS], int *solucionIdx) {
    int visitados[MAX_ESTADOS]= {0};
    int frente=0, fondo=1;
    copiarTablero(estados[0].tablero, tableroInicial);
    estados[0].pasos=0; estados[0].padre=-1;
    while(frente<fondo) {
        Estado actual=estados[frente];
        if(iguales(actual.tablero, (int (*)[N])estadoMeta)) {
            *solucionIdx=frente;
            return 1;
        }
        int filaCero, colCero;
        buscarCero(actual.tablero, &filaCero, &colCero);
        for(int m=0;m<4;m++) {
            int nf=filaCero+movimientosPosibles[m][0];
            int nc=colCero+movimientosPosibles[m][1];
            if(nf>=0 && nf<N && nc>=0 && nc<N) {
                int nuevoTablero[N][N];
                copiarTablero(nuevoTablero, actual.tablero);
                nuevoTablero[filaCero][colCero]=nuevoTablero[nf][nc];
                nuevoTablero[nf][nc]=0;
                // Verificar si ya fue visitado
                int yaVisitado=0;
                for(int k=0;k<fondo;k++)
                    if(iguales(estados[k].tablero, nuevoTablero)) { yaVisitado=1; break; }
                if(!yaVisitado && fondo<MAX_ESTADOS) {
                    copiarTablero(estados[fondo].tablero, nuevoTablero);
                    estados[fondo].pasos=actual.pasos+1;
                    estados[fondo].padre=frente;
                    fondo++;
                }
            }
        }
        frente++;
    }
    return 0;
}

// A* (búsqueda informada)
int busquedaA(int tableroInicial[N][N], Estado estados[MAX_ESTADOS], int *solucionIdx) {
    int frente=0, fondo=1;
    copiarTablero(estados[0].tablero, tableroInicial);
    estados[0].pasos=0; estados[0].padre=-1;
    estados[0].heuristica=manhattan(tableroInicial);
    while(frente<fondo) {
        // Buscar el estado con menor heuristica+pasos
        int mejor=frente;
        for(int k=frente+1;k<fondo;k++)
            if(estados[k].pasos+estados[k].heuristica < estados[mejor].pasos+estados[mejor].heuristica)
                mejor=k;
        // Intercambiar
        Estado temp=estados[frente];
        estados[frente]=estados[mejor];
        estados[mejor]=temp;
        Estado actual=estados[frente];
        if(iguales(actual.tablero, (int (*)[N])estadoMeta)) {
            *solucionIdx=frente;
            return 1;
        }
        int filaCero, colCero;
        buscarCero(actual.tablero, &filaCero, &colCero);
        for(int m=0;m<4;m++) {
            int nf=filaCero+movimientosPosibles[m][0];
            int nc=colCero+movimientosPosibles[m][1];
            if(nf>=0 && nf<N && nc>=0 && nc<N) {
                int nuevoTablero[N][N];
                copiarTablero(nuevoTablero, actual.tablero);
                nuevoTablero[filaCero][colCero]=nuevoTablero[nf][nc];
                nuevoTablero[nf][nc]=0;
                // Verificar si ya fue visitado
                int yaVisitado=0;
                for(int k=0;k<fondo;k++)
                    if(iguales(estados[k].tablero, nuevoTablero)) { yaVisitado=1; break; }
                if(!yaVisitado && fondo<MAX_ESTADOS) {
                    copiarTablero(estados[fondo].tablero, nuevoTablero);
                    estados[fondo].pasos=actual.pasos+1;
                    estados[fondo].padre=frente;
                    estados[fondo].heuristica=manhattan(nuevoTablero);
                    fondo++;
                }
            }
        }
        frente++;
    }
    return 0;
}

void imprimirSolucion(Estado estados[MAX_ESTADOS], int solucionIdx) {
    int camino[MAX_ESTADOS], n=0;
    while(solucionIdx!=-1) {
        camino[n++]=solucionIdx;
        solucionIdx=estados[solucionIdx].padre;
    }
    for(int i=n-1;i>=0;i--)
        imprimirTablero(estados[camino[i]].tablero);
}

int main() {
    int tableroInicial[N][N]={
        {5, 1, 2, 4},
        {0, 6, 3, 8},
        {9, 10, 7, 12},
        {13, 14, 11, 15}
    };
    Estado estados[MAX_ESTADOS];
    int solucionIdx;
    printf("Solución búsqueda a ciegas (puede ser lenta):\n");
    if(busquedaCiegas(tableroInicial, estados, &solucionIdx))
        imprimirSolucion(estados, solucionIdx);
    else
        printf("No se encontró solución con búsqueda a ciegas.\n");

    printf("\nSolución búsqueda informada (A*):\n");
    if(busquedaA(tableroInicial, estados, &solucionIdx))
        imprimirSolucion(estados, solucionIdx);
    else
        printf("No se encontró solución con búsqueda informada.\n");
    return 0;
}


